internal/tel/v2/tel.go

```diff
@@ -0,0 +1,350 @@
+	// LocationAlloyDBs the location of the AlloyDB instance.
+	Location string
+	// Cluster is the name of the AlloyDB cluster.
+	Cluster string
+	// Instance is the name of the AlloyDB instance.
+	Instance string
+}
+
+// MetricRecorder defines the interface for recording metrics related to the
+// internal operations of alloydbconn.Dialer.
+type MetricRecorder interface {
+	Shutdown(context.Context) error
+	RecordBytesRxCount(context.Context, int64, Attributes)
+	RecordBytesTxCount(context.Context, int64, Attributes)
+	RecordDialCount(context.Context, Attributes)
+	RecordDialLatency(context.Context, int64, Attributes)
+	RecordOpenConnection(context.Context, Attributes)
+	RecordClosedConnection(ctx context.Context, a Attributes)
+	RecordRefreshCount(ctx context.Context, a Attributes)
+}
+
+// DefaultExportInterval is the interval that the metric exporter runs. It
+// should always be 60s. This value is exposed as a var to faciliate testing.
+var DefaultExportInterval = 60 * time.Second
+
+// NewMetricRecorder creates a MetricRecorder. When the configuration is not
+// enabled, a null recorder is returned instead.
+func NewMetricRecorder(ctx context.Context, l debug.ContextLogger, cfg Config, opts ...option.ClientOption) MetricRecorder {
+	if !cfg.Enabled {
+		l.Debugf(ctx, "disabling built-in metrics")
+		return nullMetricRecorder{}
+	}
+	eopts := []cmexporter.Option{
+		cmexporter.WithCreateServiceTimeSeries(),
+		cmexporter.WithProjectID(cfg.ProjectID),
+		cmexporter.WithMonitoringClientOptions(opts...),
+		cmexporter.WithMetricDescriptorTypeFormatter(func(m metricdata.Metrics) string {
+			return "alloydb.googleapis.com/client/connector/" + m.Name
+		}),
+		cmexporter.WithMonitoredResourceDescription(monitoredResource, []string{
+			ProjectID, Location, Cluster, Instance, ClientID,
+		}),
+	}
+	exp, err := cmexporter.New(eopts...)
+	if err != nil {
+		l.Debugf(ctx, "built-in metrics exporter failed to initialize: %v", err)
+		return nullMetricRecorder{}
+	}
+
+	res := resource.NewWithAttributes(monitoredResource,
+		// The gcp.resource_type is a special attribute that the exporter
+		// transforms into the MonitoredResource field.
+		attribute.String("gcp.resource_type", monitoredResource),
+		attribute.String(ProjectID, cfg.ProjectID),
+		attribute.String(Location, cfg.Location),
+		attribute.String(Cluster, cfg.Cluster),
+		attribute.String(Instance, cfg.Instance),
+		attribute.String(ClientID, cfg.ClientID),
+	)
+	p := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(
+			exp,
+			// The periodic reader runs every 60 seconds by default, but set
+			// the value anyway to be defensive.
+			sdkmetric.WithInterval(DefaultExportInterval),
+		)),
+		sdkmetric.WithResource(res),
+	)
+	m := p.Meter(meterName, metric.WithInstrumentationVersion(cfg.Version))
+
+	mDialCount, err := m.Int64Counter(dialCount)
+	if err != nil {
+		l.Debugf(ctx, "built-in metrics exporter failed to initialize dial count metric: %v", err)
```

Comment:
Why do we not call `exp.Shutdown(context.Background())` here but do it when all of the other metrics fail to initialize?

Reply:
That's a mistake -- fixed.